### PR TITLE
ENH Make HMC and NUTS respect floatX.

### PR DESCRIPTION
**Pull request by twiecki**
The theano setting `warn_float64 = warn` was super useful for this.

Supersedes https://github.com/pymc-devs/pymc3/pull/1338.

### CODE REVIEW

**Review by springcoil — COMMENTED** [review]
LGTM.

**Comment by springcoil** [comment]
Maybe @fonnesbeck should look at this too or @ColCarroll. Or @tyarkoni

File: pymc3/step_methods/hmc/nuts.py [hunk 1]
```diff
@@ -57,15 +57,15 @@ def astep(self, q0):
...
+        u = floatX(nr.uniform())
         q = qn = qp = q0
         pn = pp = p0
 
         tree_size, depth = 1., 0
         keep_sampling = True
 
         while keep_sampling:
-            direction = bern(0.5) * 2 - 1
+            direction = floatX(bern(0.5)) * 2 - 1
```
Comment (ColCarroll) [inline pymc3/step_methods/hmc/nuts.py]: we should probably just have the function `bern` return a floatX.  Surprisingly, replacing this `bern` with calls to `numpy.random.binomial` made sampling marginally slower.

**Review by ColCarroll — APPROVED** [review]
LGTM!  Thanks for updating this

File: pymc3/step_methods/hmc/nuts.py (hunk 1 above)
Comment (twiecki) [reply pymc3/step_methods/hmc/nuts.py]: I don't think we need it at all actually.

**Comment by twiecki** [comment]
@tyarkoni Would be great if you could give this a test-drive.

**Comment by twiecki** [comment]
@Spaak too.

**Comment by ColCarroll** [comment]
In my memory, theano was upset if direction wasn't cast.  Maybe better now?

**Comment by twiecki** [comment]
> In my memory, theano was upset if direction wasn't cast.  Maybe better now?

Sorry, we don't need to cast it when calling `bern` because the variable is first used for python-indexing, we only need to cast it in leapfrog where I now do the floatX conversion.

**Comment by Spaak** [comment]
NUTS and HMC are definitely running partly on the GPU now (I'm getting GPU usage ~30% when sampling). However, something still isn't right, as GPU is about 4-5x slower for both methods than CPU (for a simple linear model with some correlated regressors). I will try to find out what's going on.

**Comment by twiecki** [comment]
Well, it's not clear that NUTS or HMC would be faster on the GPU. Or that it would depend on the model. Maybe try with this:

```python
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

np.random.seed(123)
D = 54
N = 581012
T = 100

true_beta = np.random.randn(D)

x_data = np.random.randn(N, D)
p = sigmoid(np.dot(x_data, true_beta))
y_data = np.array([np.random.binomial(1, i) for i in p])

import pymc3 as pm
import theano.tensor as tt
with pm.Model() as model:
    beta = pm.Normal('beta', mu=tt.zeros(D), sd=tt.ones(D), shape=D, 
                     testval=true_beta.astype('float32'))
    obs = pm.Bernoulli('obs', p=pm.math.sigmoid(pm.math.dot(x_data.astype('float32'), beta)), 
                       observed=y_data.astype('int32'))
    
    step = pm.HamiltonianMC(scaling={'beta': np.array([.1]*D).astype('float32')}, profile=True)
    #step = pm.NUTS(scaling={'beta': np.array([.1]*D).astype('float32')}, profile=True)
    step.step_size = .05
```

You can then analyze `step.leapfrog.profile.summary()`.

**Comment by Spaak** [comment]
Ah yes I think it will depend on the model. When I sample from a more complicated model I'm using, there is actually a significant speedup for HMC and NUTS. The example code you posted shows about a ~1.5x speedup for GPU vs CPU; for my model the speedup is bigger.

Of note might be the following top line in the profile summary for the code you posted:

```
  95.4%    95.4%      14.357s       3.88e-01s     Py      37       1   theano.scan_module.scan_op.Scan
```

so almost all the time is being spent inside the Scan op.

**Comment by twiecki** [comment]
Yeah, I noticed that too. @ColCarroll I think you converted the loop to a scan, did you check if it made thing slower?

**Comment by twiecki** [comment]
But very cool that the model is sped up! Can we somehow identify parts that are not run on the GPU?

**Comment by ColCarroll** [comment]
Thomas -- I've been benchmarking on super simple models, but I thought the
scan sped things up (on the CPU).  I'm going to add an issue to collect
some benchmarking scripts and tools so these questions are easier to answer.

**Pull request merged**